src/adam7.rs

```diff
@@ -116,49 +121,56 @@ fn expand_adam7_bits(
         5 => (4, 2, 2, 0),
         6 => (2, 0, 2, 1),
         7 => (2, 1, 1, 0),
-        _ => panic!("Adam7 pass out of range: {}", pass),
+        _ => {
+            // `InterlaceInfo.pass` is a non-`pub`lic field.  `InterlaceInfo` is expected
+            // to maintain an invariant that `pass` is valid.
+            panic!("Invalid `InterlaceInfo.pass`");
+        }
     };
 
     // the equivalent line number in progressive scan
     let prog_line = line_mul * line_no + line_off;
-    // line width is rounded up to the next byte
-    let line_width = (width * bits_pp + 7) & !7;
-    let line_start = prog_line * line_width;
-    let start = line_start + (samp_off * bits_pp);
-    let stop = line_start + (width * bits_pp);
+    let line_start = prog_line * row_stride_in_bytes * 8;
 
-    (start..stop).step_by(bits_pp * samp_mul)
+    (0..interlaced_width)
+        .map(move |i| i * samp_mul + samp_off)
+        .map(move |i| i * bits_pp)
+        .map(move |bits_offset| bits_offset + line_start)
 }
 
-/// Expands an Adam 7 pass
+/// Copies pixels from `interlaced_row` into the right location in `img`.
+///
+/// First bytes of `img` should belong to the top-left corner of the currently decoded frame.
+///
+/// `img_row_stride` specifies an offset in bytes between subsequent rows of `img`.
+/// This can be the width of the current frame being decoded, but this is not required - a bigger
+/// stride may be useful if the frame being decoded is a sub-region of `img`.
+///
+/// `interlaced_row` and `interlace_info` typically come from `Reader.next_interlaced_row`, but
+/// this is not required.  In particular, before calling `expand_interlaced_row` one may need to
+/// expand the decoded row, so that its format and `bits_per_pixel` matches that of `img`.
+/// Note that in initial Adam7 passes the `interlaced_row` may contain less pixels that the width
+/// of the frame being decoded (e.g. it contains only 1/8th of pixels in the initial pass).
 pub fn expand_pass(
     img: &mut [u8],
-    width: u32,
-    scanline: &[u8],
-    pass: u8,
-    line_no: u32,
-    bits_pp: u8,
+    img_row_stride: usize,
+    interlaced_row: &[u8],
+    interlace_info: &InterlaceInfo,
```

Comment:
Thanks for the feedback!  You have a good point.  This was also brought up in an earlier comment by @adetaylor (I probably should have gotten an internal review before submitting this PR - I am still working out the kinks in [the process I can use when working with cross-repo changes](https://docs.google.com/document/d/1RJ5IU8DiNykZkl6Epy-xKMhW3fm311m67qTdIS_S_OQ/edit?usp=sharing).

I am right now leaning toward the idea of introducing a public `Adam7Info` struct (conservatively starting with non-`pub` fields).  Postponing exposing `pass`, `line`, and `width` fields through the public API should keep this PR more focused and easier to review.

I'll try to finish this later today.

Reply:
A few other thoughts that may influence the shape of the new public API:

* `InterlaceInfo` is technically equivalent to `Option<Adam7Info>`, but I lean toward keeping a separate `InterlaceInfo` `enum`
    - Having an explicitly named type seems nice, even if we lose some of `Option<T>` APIs.
    - I am not sure if removing `InterlaceInfo` would not be considered a breaking change.  On one hand `Row`, `InterlacedRow` and `InterlaceInfo` are *not* `pub`-lically reexported or documented.  OTOH, accessing `Row.data` works fine without any `png` crate changes (storing of `InterlacedRow.interlace` doesn't work though).
* I notice that https://en.wikipedia.org/wiki/Adam7_algorithm demonstration fills more than a single row during initial adam7 passes.  I think that this is not needed for Skia/Chromium needs right now, but I haven't yet been able to try all Blink tests.  And the current API is needed by `Reader.next_frame`, so we probably should keep and expose it anyway.  In the unlikely even that this is needed, then maybe a separate public API can be added in the future: `splat_interlaced_row`?
    * I am not 100% sure if removing `InterlaceInfo` might be considered a breaking change at this point, since it is kind of exposed via `InterlacedRow.interlace` (which is [exposed via a `pub` export]() AFAICT - not sure why this is not covered by https://docs.rs/png) which gets returned from [`Reader.next_interlaced_row`](https://docs.rs/png/latest/png/struct.Reader.html#method.next_interlaced_row)

Reply:
And one more argument against replacing `InterlaceInfo` with `Option<Adam7Info>`: in theory PNG spec could adopt another interlacing scheme in the future (although this seems unlikely to me).  OTOH, if that happened, then `expand_interlaced_row` should be renamed to `expand_adam7_row`... :-/

At any rate, I refreshed the PR and now the API should by construction disallow panics (i.e. there is now no way to pass `InterlaceInfo::Null` to `fn expand_interlaced_row`).  So, let me mark this review thread as resolved.